contracts/staking/test/unit_tests/mixin_scheduler_test.ts

```diff
@@ -0,0 +1,120 @@
+            const initCurrentEpochStartTimeInSeconds = new BigNumber(10);
+            const tx = testContract.initMixinSchedulerTest.awaitTransactionSuccessAsync(
+                initCurrentEpochStartTimeInSeconds,
+            );
+            return expect(tx).to.revertWith(
+                new StakingRevertErrors.InitializationError(
+                    StakingRevertErrors.InitializationErrorCodes.MixinSchedulerAlreadyInitialized,
+                ),
+            );
+        });
+    });
+
+    describe('_goToNextEpoch', () => {
+        it('Should succeed if epoch end time is strictly less than to block timestamp', async () => {
+            const epochEndTimeDelta = new BigNumber(-10);
+            const txReceipt = await testContract.goToNextEpochTest.awaitTransactionSuccessAsync(epochEndTimeDelta);
+            // tslint:disable-next-line no-unnecessary-type-assertion
+            const testLog: TestMixinSchedulerGoToNextEpochTestInfoEventArgs = (txReceipt.logs[0] as LogWithDecodedArgs<
+                TestMixinSchedulerGoToNextEpochTestInfoEventArgs
+            >).args;
+            const currentEpoch = await testContract.currentEpoch.callAsync();
+            const currentEpochStartTimeInSeconds = await testContract.currentEpochStartTimeInSeconds.callAsync();
+            expect(currentEpoch).to.bignumber.equal(testLog.oldEpoch.plus(1));
+            expect(currentEpochStartTimeInSeconds).to.bignumber.equal(testLog.blockTimestamp);
+        });
+
+        it('Should succeed if epoch end time is equal to block timestamp', async () => {
+            const epochEndTimeDelta = new BigNumber(0);
+            const txReceipt = await testContract.goToNextEpochTest.awaitTransactionSuccessAsync(epochEndTimeDelta);
+            // tslint:disable-next-line no-unnecessary-type-assertion
+            const testLog: TestMixinSchedulerGoToNextEpochTestInfoEventArgs = (txReceipt.logs[0] as LogWithDecodedArgs<
+                TestMixinSchedulerGoToNextEpochTestInfoEventArgs
+            >).args;
+            const currentEpoch = await testContract.currentEpoch.callAsync();
+            const currentEpochStartTimeInSeconds = await testContract.currentEpochStartTimeInSeconds.callAsync();
+            expect(currentEpoch).to.bignumber.equal(testLog.oldEpoch.plus(1));
+            expect(currentEpochStartTimeInSeconds).to.bignumber.equal(testLog.blockTimestamp);
+        });
+
+        it('Should revert if epoch end time is strictly greater than block timestamp', async () => {
+            const epochEndTimeDelta = new BigNumber(10);
+            const tx = testContract.goToNextEpochTest.awaitTransactionSuccessAsync(epochEndTimeDelta);
+            try {
+                await tx;
+                // tslint:disable-next-line no-empty
+            } catch (e) {}
+
+            // Mine the block that this tx would've been in.
+            await env.web3Wrapper.mineBlockAsync();
```

Comment:
This is a pretty cool pattern. Might make sense to make a function and put in test-utils, but optional for this PR.

Reply:
🙏 - although it's such an edge case idk if it would be worth abstracting as-is.

Reply:
Fair enough lol